api_etl/adapters/ubr_adapter.py

```diff
@@ -100,3 +101,115 @@ def parse_individual_role(member_dict):
+            if location_data:
+                result.append(location_data)
+
+        return result
+
+    def _process_district(self, row: dict, location_cache: dict) -> dict:
+        geo_location_code = row.get("geo_location_code")
+        geo_location_name = row.get("geo_location_name")
+
+        if not geo_location_code or not geo_location_name:
+            logger.warning("Skipping district due to missing geo_location_code or geo_location_name")
+            return None
+
+        # Determine the region code from the first character of the geo_location_code
+        region_code = geo_location_code[0]
+
+        region = self._get_or_cache_location(region_code, "R", location_cache)
+        if not region:
+            logger.error(f"Region with code {region_code} not found for district {geo_location_code}")
+            return None
+
+        return {
+            "code": geo_location_code,
+            "name": geo_location_name,
+            "type": "D",
+            "parent": region,
+        }
+
+    def _process_ta(self, row: dict, location_cache: dict) -> dict:
+        geo_location_code = row.get("geo_location_code")
+        geo_location_name = row.get("geo_location_name")
+        parent_geo_location_code = row.get("parent_geo_location_code")
+
+        if not geo_location_code or not geo_location_name or not parent_geo_location_code:
+            logger.warning("Skipping TA due to missing geo_location_code, geo_location_name, or parent_geo_location_code")
+            return None
+
+        parent_district = self._get_or_cache_location(parent_geo_location_code, "D", location_cache)
+        if not parent_district:
+            logger.error(f"Parent district with code {parent_geo_location_code} not found for TA {geo_location_code}")
+            return None
+
+        return {
+            "code": geo_location_code,
+            "name": geo_location_name,
+            "type": "W",
+            "parent": parent_district,
+        }
+
+    def _process_village(self, row: dict, location_cache: dict) -> dict:
+        geo_location_code = row.get("geo_location_code")
+        geo_location_name = row.get("geo_location_name")
+        parent_geo_location_code = row.get("parent_geo_location_code")
+
+        if not geo_location_code or not geo_location_name or not parent_geo_location_code:
+            logger.warning("Skipping Village due to missing geo_location_code, geo_location_name, or parent_geo_location_code")
+            return None
+
+        parent_geo_location_code = parent_geo_location_code[:5]
+
+        parent_ta = self._get_or_cache_location(parent_geo_location_code, "W", location_cache)
+        if not parent_ta:
+            logger.error(f"Parent TA with code {parent_geo_location_code} not found for Village {geo_location_code}")
+            return None
+
+        return {
+            "code": geo_location_code,
+            "name": geo_location_name,
+            "type": "V",
+            "parent": parent_ta,
+        }
```

Comment:
Similarly to logic repetitions in urb_source, here consider extract shared logic from _process_village, _process_ta, and _process_district to keep code DRY.